In [1]:
from os import listdir,mkdir
from os.path import isfile, join, isdir,exists
import pandas as pd
import numpy as np
from scipy import stats
import re
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
from myplots import roundup, rounddown, find_decimal_fold, percentile_cut_off, rarefaction_calc, rarefaction_plot,draw_correlation_scatter
from matplotlib.ticker import FormatStrFormatter
import cPickle as pickle
from Bio.SeqUtils import GC
import seaborn as sns
import random
from scipy.stats import pearsonr
from skbio.diversity.alpha import shannon, simpson, berger_parker_d

from pop_organize import get_sample_data, get_sample_with_dfs
from SufficientStatistics import *
from MyFunctionsShani import *


done1


In [2]:
import time
cdate=str(time.strftime("%d%m%Y"))
cdate

'24122017'

## generate list of all uniqueAAs per sample in Adaptive cohort:

#### define the function:

In [7]:
#this function is copied from the notebook 'Public analysis- PNP cohort':


def gen_uniqueAAdf(sample_name,data_folder):
    file1='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/uniqueAAdfs/%s' %(data_folder,sample_name) 
    files=[f for f in listdir(uniqueAAdfs_folder) if isfile(join(uniqueAAdfs_folder, f))]


    if sample_name not in files:
        data_folder='TCR_demo_data'
        sample_df=pd.read_table("/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/SamplesForAnalysis/%s.tsv" %(data_folder,sample_name))
        sample_df=sample_df.rename(columns={'count (templates/reads)':'count (templates)','count (reads)':'count (templates)'})
        sample_df['prod_stat']=np.where(sample_df['sequenceStatus'] == 'In',1,0)
        f={'count (templates)':['sum','count'],'prod_stat':'mean'}
        uniqAA=sample_df.groupby('aminoAcid').agg(f)
        uniqAA=uniqAA.drop(('count (templates)','count'),axis=1)
        uniqAA['Sample']=sample_name
        uniqAA.to_pickle(file1)
    else:
        print('found uniqueAAdf for this sample...')



#### get list of adaptive samples:

In [8]:
#get list of adaptive samples:

data_folder='TCR_demo_data'
sample_dfs_folder='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/SamplesForAnalysis' %data_folder
sample_filenames = [f for f in listdir(sample_dfs_folder) if isfile(join(sample_dfs_folder, f))]
sample_filenames=[f.strip('.tsv') for f in sample_filenames]
print len(sample_filenames)


788


#### generate the list:

In [9]:


uniqueAAdfs_folder='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/uniqueAAdfs/' %data_folder
if not isdir(uniqueAAdfs_folder):
    mkdir(uniqueAAdfs_folder)

for n,sample_name in enumerate(sample_filenames):    
        print  n,sample_name
        gen_uniqueAAdf(sample_name,data_folder)

0 HIP02877
1 HIP10602
2 HIP13376
3 HIP01266
4 HIP01499
5 Keck0002_MC1
6 HIP08710
7 HIP09150
8 Keck0006_MC1


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (9,24) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


9 HIP14016
10 HIP13956
11 HIP01393
12 HIP10726
13 HIP13252
14 HIP13987
15 HIP01055
16 HIP13809
17 HIP09364
18 HIP13951


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


19 HIP13284
20 HIP13142
21 HIP13193
22 HIP13710
23 HIP05409
24 HIP03184
25 HIP13875
26 HIP16738
27 HIP13699
28 HIP08400
29 HIP02928
30 HIP05551
31 HIP08989
32 HIP03597
33 HIP13303
34 HIP03651
35 Keck0102_MC1
36 Keck0025_MC1
37 HIP14196
38 Keck0021_MC1
39 HIP09001
40 HIP13923
41 Keck0106_MC1
42 HIP13227
43 HIP09159
44 HIP12538
45 Keck0029_MC1
46 HIP14064
47 HIP03004
48 HIP10443
49 HIP13220
50 HIP02805
51 HIP13800
52 HIP14018
53 HIP00826


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


54 HIP13958
55 HIP09235
56 HIP14140
57 HIP17723
58 HIP05524
59 HIP09122
60 HIP13989
61 HIP14051
62 HIP14080
63 HIP10639
64 HIP00594
65 HIP12034
66 HIP19089
67 HIP14109
68 HIP13911
69 HIP10694
70 Keck0048_MC1
71 HIP00813
72 HIP12099
73 HIP14175
74 HIP11774
75 HIP04578
76 HIP13786
77 HIP17698
78 HIP00771
79 HIP02024
80 HIP13757
81 Keck0040_MC1
82 HIP17370
83 HIP14172
84 HIP13518
85 HIP03618
86 Keck0044_MC1
87 HIP13916
88 HIP04576
89 HIP14024
90 HIP05757
91 HIP09097
92 HIP14494
93 HIP09046
94 HIP05832
95 HIP03732
96 HIP13964
97 HIP12165
98 HIP14089
99 HIP14217
100 HIP01253
101 HIP13847
102 HIP13722
103 Keck0067_MC1
104 HIP08725
105 HIP05563
106 HIP04455
107 HIP14107
108 Keck0097_MC1
109 HIP05331
110 HIP13449
111 HIP09041
112 Keck0093_MC1
113 HIP09119
114 HIP03385
115 Keck0063_MC1
116 HIP12097
117 HIP13511
118 Keck0055_MC1


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (9,11,24) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


119 HIP17462
120 HIP05665
121 HIP13887
122 HIP13352
123 HIP00715
124 HIP13856
125 HIP13383
126 HIP13972
127 HIP13276
128 HIP00985


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (9,10,11,13,18,24) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


129 HIP01129
130 HIP13489
131 HIP00769
132 Keck0051_MC1
133 HIP08507
134 HIP13975
135 HIP01865
136 HIP13610
137 Keck0059_MC1
138 HIP08972
139 HIP14206
140 HIP13355
141 HIP13880
142 HIP11717
143 Keck0072_MC1
144 HIP13168
145 HIP11989
146 Keck0082_MC1
147 HIP13746
148 HIP00805
149 HIP13823
150 HIP01313
151 Keck0086_MC1
152 HIP01004
153 HIP01465
154 Keck0076_MC1
155 HIP14096
156 HIP00926
157 HIP14118
158 HIP13900
159 HIP10815
160 HIP09022
161 HIP04605
162 HIP17534
163 HIP12088
164 HIP13741
165 HIP01597
166 HIP13935
167 HIP10820
168 HIP04527
169 HIP08236
170 HIP05535
171 HIP10358
172 HIP11518
173 HIP10389
174 HIP13949
175 HIP08816
176 HIP13463
177 HIP14009
178 HIP13774
179 HIP14187
180 HIP14241
181 HIP14156
182 Keck0017_MC1
183 HIP13773
184 HIP13932
185 HIP13686
186 Keck0013_MC1
187 HIP14072
188 HIP08596
189 HIP10730
190 HIP13625
191 HIP01850
192 HIP13244
193 HIP13818
194 HIP14000
195 HIP04552
196 HIP09062
197 HIP05540
198 HIP01798
199 HIP14124
200 Keck0038_MC1
201 HIP08598
202 HIP13360
20

730 HIP04532
731 HIP14138
732 HIP13920
733 HIP10447
734 HIP13831
735 HIP03111
736 HIP01582
737 HIP13754
738 HIP02780
739 Keck0119_MC1
740 HIP15854
741 HIP14055
742 HIP02848
743 HIP01805
744 HIP02103
745 HIP13567
746 Keck0115_MC1
747 HIP01470
748 Keck0032_MC1
749 HIP00934
750 Keck0036_MC1
751 HIP13782
752 HIP05838
753 HIP13753
754 HIP00775
755 HIP12900
756 HIP05960
757 HIP00999
758 Keck0111_MC1
759 HIP14176
760 HIP03505
761 HIP00707
762 Keck0019_MC1
763 HIP14213
764 HIP14020
765 HIP05444
766 HIP17585
767 HIP01870
768 HIP00997
769 HIP14178
770 HIP00819
771 HIP02962
772 HIP10376
773 Keck0015_MC1
774 HIP13263
775 HIP13515
776 HIP13967
777 HIP03381
778 HIP13396
779 Keck0011_MC1
780 HIP04509
781 HIP03502
782 HIP05398
783 HIP01140
784 HIP17760
785 HIP04480
786 HIP14214


/usr/wisdom/python/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (1,9,10,11,18,24) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


787 HIP14103


## define a function that calculates the percent of shared sequences in a dataset, given the sample names, number of samples to select, number of sequences to select from each sample, and number of calculation repeats. 

(1) choose randomly nSamples
(2) for each sample, load its uniqueAA df, and randomly choose nSequences, and add them to the new uniqueAA df
(3) count values for the new uniqueAA sequences
(4) calculate percent of public and save value
(5) repeat for nRepeats and calculate the mean.



In [63]:
#this function calculates the percent of shared sequences in a dataset, given the 

def testPublicRatesInSelectedSamples(sample_filenames,nSamples,nSequences,nRepeats,useSequenceWeights):
    
    percPublicDF=pd.DataFrame()
    for r in range(nRepeats):
        print 'starting repeat %s...' %r
        
        # (1) choose randomly 'nSamples' samples (currently nSamples=434)
        print 'choosing %s samples...' %nSamples
        selectedSamples=np.random.choice(sample_filenames, nSamples, replace=False)

        #(2) for each sample, load its uniqueAA df, and randomly choose 'nSequences' sequences, and add them to the new uniqueAA df
        # (currently nSequences=12,756)
        print 'loading uniqueAA df selecting %s sequences and adding to the large uniqueAA df...' %nSequences
        TotaluniqueAA=pd.DataFrame()
        for n, sample_name in enumerate(selectedSamples):
                print n, sample_name
                file1='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/uniqueAAdfs/%s' %(data_folder,sample_name)
                uniqueAA=pd.read_pickle(file1)
                uniqueAAlist=list(uniqueAA.index)
                if useSequenceWeights:
                    countsList=list(uniqueAA[('count (templates)','sum')])
                    countsListSum=np.sum(countsList)
                    countsListNorm=[float(x)/countsListSum for x in countsList]
                else:
                    countsListNorm=None
                if len(uniqueAAlist)>nSequences:
                    chosenUniqueAAs= np.random.choice(uniqueAAlist, nSequences, replace=False,p=countsListNorm)
                else:
                    chosenUniqueAAs= np.random.choice(uniqueAAlist, len(uniqueAAlist), replace=False,p=countsListNorm)
                    print '% seqeunces only were taken from sample %s' %(len(uniqueAAlist),sample_name)
                chosenUniqueAAsDF=pd.DataFrame({'aminoAcid':chosenUniqueAAs,'Sample': sample_name})
                TotaluniqueAA=pd.concat([TotaluniqueAA,chosenUniqueAAsDF])
        
        #(3) count values for the new uniqueAA sequences:
        print 'counting values for uniqueAA sequences in repeat %s' %r
        TotaluniqueAA=TotaluniqueAA.set_index('aminoAcid')
        CountSharing=TotaluniqueAA.index.value_counts()
        
        
        #(4) calculate percent of public and save value:
        print 'calculating percent of public sequences and saving value for repeat %s' %r
        CountSharing=pd.DataFrame(CountSharing)
        CountSharing=CountSharing.rename(columns={'aminoAcid':'n_shared_samples'})
        Npublic=len(CountSharing[CountSharing['n_shared_samples']>1])
        Nprivate=len(CountSharing[CountSharing['n_shared_samples']==1])
        percPublic=float(Npublic)*100/Nprivate
        
        print 'percPublic in repeat %s is %s' %(r,percPublic)
        percPublicDF.loc[r,'percPublic']=percPublic
        
    return percPublicDF
    
        
        
    

## calculate percPublic for 434 samples, 12576 sequences per sample, with 10 calculation repeats:

In [60]:
sample_filenames=sample_filenames
nSamples=434
nSequences=12576
nRepeats=10

percPublicDF=testPublicRatesInSelectedSamples(sample_filenames,nSamples,nSequences,nRepeats)
file1='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_real_data/PublicSeqAnalysis/percPublicCalc_AdaptiveDataSimulation_%s_%s_%s' %(nSamples,nSequences,
                                                                                                                                  nRepeats)
percPublicDF.to_pickle(file1)

starting repeat 0...
choosing 434 samples...
loading uniqueAA df selecting 12576 sequences and adding to the large uniqueAA df...
0 HIP13814
1 HIP05331
2 HIP14045
3 HIP13414
4 HIP05390
5 HIP08230
6 HIP13877
7 HIP09430
8 HIP11711
9 HIP11758
10 HIP09366
11 HIP14231
12 HIP01499
13 Keck0062_MC1
14 HIP13178
15 HIP13513
16 HIP13782
17 HIP05763
1084eqeunces only were taken from sample HIP05763
18 Keck0065_MC1
19 HIP08499
20 HIP13658
21 HIP14240
22 HIP00822
23 HIP13920
24 HIP01218
25 HIP04611
26 HIP08507
27 Keck0023_MC1
28 HIP13916
29 HIP09046
30 HIP15855
31 HIP05595
32 HIP15861
33 HIP14227
34 Keck0109_MC1
35 HIP00904
36 HIP11613
37 HIP04471
38 HIP01197
39 HIP13986
40 HIP03381
41 HIP13183
42 HIP03618
43 HIP17534
44 HIP14095
45 HIP13887
46 HIP03814
47 HIP10568
48 HIP03383
49 HIP08200
50 HIP10424
51 Keck0067_MC1
52 HIP10823
53 HIP14074
54 HIP13361
55 Keck0101_MC1
56 HIP00715
57 HIP13661
58 HIP03233
59 HIP08710
60 HIP04552
61 Keck0093_MC1
62 Keck0066_MC1
63 Keck0049_MC1
64 HIP13671
65 HIP13636
66

147 HIP13274
148 HIP04527
149 HIP13809
150 HIP10545
151 HIP11518
152 HIP14060
153 HIP14178
154 Keck0033_MC1
155 Keck0038_MC1
156 Keck0075_MC1
157 Keck0076_MC1
158 HIP13722
159 HIP13771
160 HIP12088
161 HIP13773
162 HIP00971
163 Keck0041_MC1
164 HIP13916
165 Keck0078_MC1
166 HIP13265
167 HIP12097
168 HIP13967
169 HIP11857
170 HIP13244
171 HIP13945
172 HIP10820
173 HIP18119
174 Keck0031_MC1
175 HIP13193
176 HIP13554
177 HIP03381
178 HIP10443
179 HIP13661
180 HIP11711
181 HIP14134
182 HIP10730
183 HIP00822
184 HIP01947
185 Keck0018_MC1
186 HIP14074
187 Keck0073_MC1
188 HIP03693
189 HIP03484
190 HIP13168
191 HIP09022
192 Keck0082_MC1
193 HIP04471
194 HIP03228
195 HIP08345
196 HIP13887
197 HIP03383
198 HIP11717
199 Keck0056_MC1
200 HIP00775
201 HIP09041
202 HIP14071
203 Keck0007_MC1
204 HIP14004
205 HIP13983
206 HIP08598
207 HIP14092
208 HIP14024
209 Keck0061_MC1
210 Keck0032_MC1
211 HIP14007
212 HIP01392
213 HIP13902
214 HIP15860
215 HIP05398
216 Keck0043_MC1
217 HIP05563
218 Keck0111_MC1


292 HIP03184
293 HIP04471
294 HIP13777
295 HIP13964
296 HIP10514
297 HIP13983
298 HIP01392
299 HIP01206
300 HIP00926
301 HIP13625
302 HIP14202
303 HIP01140
304 HIP14148
305 HIP11649
306 HIP09029
307 HIP14041
308 Keck0065_MC1
309 HIP13929
310 HIP14238
311 HIP13803
312 HIP14045
313 HIP13774
314 HIP01091
315 Keck0105_MC1
316 HIP13880
317 HIP05960
318 HIP13966
319 Keck0013_MC1
320 HIP13663
321 Keck0054_MC1
322 Keck0052_MC1
323 HIP05948
324 Keck0070_MC1
325 HIP12900
326 HIP05311
327 HIP11613
328 HIP00869
329 HIP14020
330 HIP13831
331 HIP05763
1084eqeunces only were taken from sample HIP05763
332 HIP13877
333 HIP05838
334 HIP15854
335 HIP13944
336 Keck0037_MC1
337 HIP14036
338 HIP17370
339 HIP13751
340 HIP09681
341 HIP00710
342 Keck0079_MC1
343 HIP05405
344 HIP12165
345 HIP08399
346 HIP01856
347 HIP02078
348 HIP14079
349 HIP13244
350 HIP11857
351 Keck0027_MC1
352 Keck0108_MC1
353 HIP04545
354 HIP03484
355 HIP12534
356 HIP13176
357 HIP10820
358 HIP13427
359 HIP13396
360 HIP13489
361 HIP04464


430 HIP13554
431 HIP01253
432 Keck0078_MC1
433 HIP01140
counting values for uniqueAA sequences in repeat 3
calculating percent of public sequences and saving value for repeat 3
percPublic in repeat 3 is 11.8814152821
starting repeat 4...
choosing 434 samples...
loading uniqueAA df selecting 12576 sequences and adding to the large uniqueAA df...
0 HIP03495
1 HIP14074
2 Keck0109_MC1
3 HIP01264
4 HIP13812
5 Keck0004_MC1
6 HIP13760
7 HIP13976
8 HIP14157
9 HIP12980
10 HIP02962
11 HIP11774
12 HIP09789
13 HIP13157
14 HIP01218
15 HIP05934
16 HIP14028
17 HIP14226
18 HIP10358
19 HIP13958
20 HIP14071
21 Keck0078_MC1
22 HIP10726
23 HIP10480
24 HIP13949
25 HIP00869
26 HIP14106
10765eqeunces only were taken from sample HIP14106
27 HIP13554
28 Keck0017_MC1
29 HIP14039
30 HIP13505
31 HIP13981
32 Keck0024_MC1
33 HIP14136
34 HIP13370
35 HIP13396
36 HIP05578
37 HIP01947
38 HIP13686
39 HIP03233
40 HIP03494
41 HIP13122
42 HIP00777
43 HIP05561
44 HIP03004
45 Keck0031_MC1
46 HIP02855
47 HIP13871
48 HIP05574


126 HIP14241
127 HIP01162
128 HIP03216
129 HIP03004
130 Keck0062_MC1
131 HIP14143
132 HIP12743
133 HIP13941
134 HIP11717
135 HIP13513
136 HIP10389
137 HIP13789
138 HIP04578
139 HIP13230
140 HIP01805
141 HIP14194
142 HIP14161
143 HIP14015
144 HIP13902
145 HIP05960
146 HIP03630
147 Keck0038_MC1
148 HIP14037
149 HIP11518
150 HIP03746
151 HIP01865
152 Keck0039_MC1
153 Keck0102_MC1
154 HIP13769
155 HIP13978
156 Keck0119_MC1
157 HIP13168
158 HIP13703
159 HIP13954
160 HIP07754
161 HIP13916
162 HIP11784
163 HIP10377
164 HIP13853
165 HIP08888
166 HIP01597
167 HIP13303
168 HIP01004
169 HIP13877
170 HIP05388
171 HIP13473
172 HIP00898
173 HIP09051
174 HIP08200
175 HIP14494
176 HIP01255
177 HIP14066
178 HIP14227
179 HIP14172
180 HIP08400
181 HIP14218
182 HIP13722
183 HIP02734
184 HIP14053
185 HIP14174
186 HIP13554
187 HIP10817
188 HIP13511
189 HIP13753
6617eqeunces only were taken from sample HIP13753
190 HIP00110
191 HIP13122
192 Keck0025_MC1
193 HIP14090
194 Keck0076_MC1
195 HIP05311
196 HIP14148

270 Keck0060_MC1
271 HIP03484
272 HIP02742
273 Keck0034_MC1
274 Keck0111_MC1
275 HIP13981
276 HIP13809
277 HIP13945
278 HIP17845
279 HIP05533
280 Keck0049_MC1
281 Keck0039_MC1
282 HIP14209
283 HIP14129
284 HIP03495
285 HIP16738
286 HIP04455
287 HIP13122
288 HIP10815
289 HIP08499
290 HIP00805
291 HIP08821
292 Keck0004_MC1
293 HIP08890
294 HIP13989
295 HIP09624
296 HIP05434
297 Keck0012_MC1
298 HIP05388
299 HIP03592
300 Keck0005_MC1
301 HIP14363
302 HIP01206
303 HIP17737
304 Keck0026_MC1
305 HIP13703
306 HIP13214
307 HIP13206
308 HIP18119
309 HIP13511
310 HIP01181
311 HIP13848
312 HIP13376
313 HIP14053
314 Keck0096_MC1
315 HIP01232
316 HIP08711
317 HIP00951
318 HIP13786
319 HIP09118
320 HIP01797
321 HIP13749
322 HIP12980
323 HIP13856
324 HIP14194
325 HIP14140
326 HIP14226
327 HIP00602
328 HIP00773
329 HIP09020
330 HIP09789
331 HIP01219
332 Keck0070_MC1
333 Keck0064_MC1
334 HIP05551
335 HIP05763
1084eqeunces only were taken from sample HIP05763
336 HIP08521
337 HIP14074
338 HIP17657
339 H

415 HIP05942
416 HIP01206
417 HIP01022
418 HIP13636
419 HIP09097
420 Keck0022_MC1
421 HIP12099
422 HIP09190
423 HIP00851
424 HIP05331
425 Keck0088_MC1
426 HIP17723
427 HIP14092
428 HIP09026
429 HIP14140
430 HIP13244
431 HIP14059
432 HIP09029
433 HIP04509
counting values for uniqueAA sequences in repeat 7
calculating percent of public sequences and saving value for repeat 7
percPublic in repeat 7 is 11.8793755484
starting repeat 8...
choosing 434 samples...
loading uniqueAA df selecting 12576 sequences and adding to the large uniqueAA df...
0 HIP13363
1 HIP02811
2 HIP09118
3 HIP02271
4 HIP01867
5 HIP13822
6 HIP03494
7 Keck0047_MC1
8 HIP17454
9 HIP08702
10 HIP05757
11 Keck0046_MC1
12 HIP00773
13 HIP09119
14 HIP13473
15 HIP05942
16 HIP11711
17 HIP03678
18 HIP11553
19 HIP13319
20 Keck0012_MC1
21 HIP14244
22 HIP14064
23 HIP04611
24 HIP00813
25 HIP05409
26 HIP08711
27 HIP13736
28 HIP00771
29 HIP08499
30 HIP17760
31 HIP01180
32 HIP14096
33 HIP14037
34 HIP03505
35 HIP16515
36 HIP02742
37 HIP03

116 Keck0014_MC1
117 Keck0018_MC1
118 HIP11058
119 Keck0026_MC1
120 HIP05841
121 HIP14205
122 HIP01836
123 HIP13015
124 HIP13695
125 HIP13720
126 HIP08711
127 HIP17760
128 HIP05941
129 Keck0082_MC1
130 HIP14136
131 HIP12743
132 HIP03502
133 HIP15861
134 HIP17698
135 Keck0102_MC1
136 HIP14237
137 HIP14134
138 HIP14074
139 HIP13376
140 HIP13671
141 HIP13350
142 Keck0006_MC1
143 HIP02112
144 HIP09046
145 Keck0011_MC1
146 HIP01797
147 HIP01091
148 HIP13902
149 HIP14494
150 HIP01223
151 HIP00999
152 HIP11717
153 Keck0021_MC1
154 HIP03383
155 HIP05331
156 HIP14000
157 HIP08702
158 HIP14092
159 HIP13319
160 HIP03695
161 HIP13214
162 HIP13741
163 HIP13284
164 HIP14178
165 Keck0111_MC1
166 HIP03381
167 Keck0015_MC1
168 HIP13306
169 HIP12129
170 HIP05388
171 HIP13636
172 HIP14230
173 HIP13168
174 HIP10597
175 HIP00775
176 HIP09190
177 HIP03495
178 HIP10821
179 HIP01596
180 HIP08888
181 HIP13823
182 HIP02103
183 HIP05467
184 HIP13949
185 Keck0003_MC1
186 HIP14223
187 HIP03812
188 HIP03484
189 HIP

In [61]:
percPublicDF

percPublic
0   11.841296
1   11.886915
2   11.875448
3   11.881415
4   11.841625
5   11.892067
6   11.907907
7   11.879376
8   11.865832
9   11.867599

In [62]:
percPublicDF.mean()

percPublic    11.873948
dtype: float64

When randomly choosing 434 samples from the adaptive cohort, and sampling 12,576 aa sequences from each (unless there are less sequences in the sample and then we take all), the mean percentage of public sequences is 11.87%. 
note that the sequence sampling was done based on template counts (the more clonal the sequence, the higher chances to sample it).
I'll run it again without taking into acount the sequence clonality:

## calculate percPublic for 434 samples, 12576 sequences per sample, with 5 calculation repeats ***with no weighted probabilitis for sequence sampling***:

In [ ]:
sample_filenames=sample_filenames
nSamples=434
nSequences=12576
nRepeats=5
useSequenceWeights=False

percPublicDF_10_5000_3_false=testPublicRatesInSelectedSamples(sample_filenames,nSamples,nSequences,nRepeats,useSequenceWeights)
file1='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_real_data/PublicSeqAnalysis/percPublicCalc_AdaptiveDataSimulation_%s_%s_%s_%s' %(nSamples,nSequences,
                                                                                                                                  nRepeats,useSequenceWeights)
percPublicDF_434_12576_5_false.to_pickle(file1)
print percPublicDF_434_12576_5_false.mean()

In [69]:
percPublicDF_434_12576_5_false=percPublicDF_10_5000_3_false
file1='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_real_data/PublicSeqAnalysis/percPublicCalc_AdaptiveDataSimulation_%s_%s_%s_%s' %(nSamples,nSequences,
                                                                                                                                  nRepeats,useSequenceWeights)
percPublicDF_434_12576_5_false.to_pickle(file1)
print percPublicDF_434_12576_5_false.mean()

percPublic    10.802665
dtype: float64


In [70]:
percPublicDF_434_12576_5_false

percPublic
0   10.795614
1   10.790761
2   10.792802
3   10.822028
4   10.812119

## look for sequences shared between datasets

In [71]:
#(6) will take a lot of time!

print 'concatenating...'
uniqueAAdfs_folder='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/%s/uniqueAAdfs/' %data_folder 
All788samples_UniqueAAdfs=concat_summarizing_dfs(uniqueAAdfs_folder)


concatenating...
number of dfs in directory: 788
the length of df list is 788


In [72]:
All788samples_UniqueAAdfs.head()

count (templates) prod_stat    Sample
                             sum      mean          
aminoAcid                                           
*A*RVWIAKNIQYF                 6         0  HIP14092
*ASRLGGGYTF                    4         0  HIP14092
*ASSGFGKLFF                    2         0  HIP14092
*ASSLAGGRTF                    4         0  HIP14092
*ASSQGSEAFF                    2         0  HIP14092

In [73]:
print 'counting...'
countSharedAdaptive788samples= All788samples_UniqueAAdfs.index.value_counts()

counting...


In [75]:
print 'saving...'
file2='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_demo_data/public_analysis/countSharedAdaptive788samples' 
countSharedAdaptive788samples.to_pickle(file2)

saving...


In [4]:
file3='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_real_data/PublicSeqAnalysis/countSharedPNP434samples' 
countSharedPNP434samples=pd.read_pickle(file3)



In [6]:
file2='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_demo_data/public_analysis/countSharedAdaptive788samples' 
countSharedAdaptive788samples=pd.read_pickle(file2)

In [7]:
countSharedPNP434samples=pd.DataFrame(countSharedPNP434samples)

In [9]:
countSharedAdaptive788samples=pd.DataFrame(countSharedAdaptive788samples)

In [10]:
sharedsequencesBetweenCohorts=pd.merge(countSharedPNP434samples,countSharedAdaptive788samples,
                                       how='inner',left_index=True, right_index=True)

sharedsequencesBetweenCohorts

aminoAcid_x  aminoAcid_y
CASSLGETQYF               278          787
CASSLGYEQYF               235          781
CASSLQETQYF               233          783
CASSLGGNTEAFF             228          783
CASSSSYEQYF               218          781
CASSLSYEQYF               217          777
CASSLNTEAFF               198          779
CASSLEETQYF               197          781
CASSLGGYEQYF              194          772
CASSLRETQYF               191          774
CASSLGDTQYF               186          772
CASSPSTDTQYF              183          779
CASSLNTGELFF              181          771
CASSLTDTQYF               181          772
CASSPSYEQYF               179          771
CASSSQETQYF               178          781
CASSLSTDTQYF              176          780
CASSDRDTGELFF             174          649
CASSLGTDTQYF              173          775
CASSFQETQYF               169          775
CASSSTDTQYF               167          771
CASSLAYEQYF               166          767
CASSSYNEQFF               164          763
CASSLSSYEQYF              164          768
CASSDSSGANVLTF            164          626
CASSLGSYEQYF              163          773
CASSLQGNTEAFF             161          768
CASSPQETQYF               160          769
CASSLGQGYEQYF             159          766
CASSLGNTEAFF              158          778
...                       ...          ...
CSVRDSSNYGYTF               1           12
CSAARGTQYF                  1           15
CASSPLDSPTNEKLFF            1            3
CASSLGRQDEQFF               1           16
CASSEVTSVSGELFF             1            1
CSGTSGKETQYF                1           10
CASSPPAVRDTQYF              1            2
CASSPVRDSYYNEQFF            1            1
CSALRTGVSYEQYF              1            4
CASSRGQGGDSPLHF             1           15
CASSRRATNEQFF               1            7
CASSMQRNTEAFF               1           18
CASSPLLGNIQYF               1            6
CASSSLSGTANEKLFF            1            1
CASSQGDSGNYEQYF             1            2
CSVALGQNTEAFF               1            9
CASSFGPGTNTEAFF             1           10
CASSLPGQQGYTF               1            3
CASGVGNEKLFF                1           14
CASSQAADSTDTQYF             1            8
CASSRAQVNTEAFF              1            9
CASSLEVSTNEKLFF             1           12
CASSLSGGAVGDTQYF            1            5
CASSEALGGRTYGYTF            1            1
CASRARDYGYTF                1           27
CASTSGGGAVDTQYF             1            4
CATSDLQQETQYF               1           11
CASSIRGLAGELFF              1            9
CSGLAGPTYEQYF               1            3
CASSQDLTQGRGYTF             1            3

[1934732 rows x 2 columns]

In [18]:
sharedsequencesBetweenCohorts=sharedsequencesBetweenCohorts.rename(columns={'aminoAcid_x':'nSamples_PNP434','aminoAcid_Y':'nSamples_Adaptive787'})

In [19]:
file4='/net/mraid08/export/genie/Lab/Personal/ShaniBAF/TCR_real_data/PublicSeqAnalysis/sharedsequencesBetweenCohorts' 
sharedsequencesBetweenCohorts.to_pickle(file4)

 #### DO NOT SAVE AS EXCEL FILE AS IT IS TOO LARGE!###

In [12]:
print len(countSharedPNP434samples)
print len(countSharedAdaptive788samples)
print len(sharedsequencesBetweenCohorts)

4244136
77465769
1934732


In [13]:
#of the PNP434 sequences - 45.6% are shared with the Adaptive787 sequences:
float(len(sharedsequencesBetweenCohorts))*100/len(countSharedPNP434samples)

45.58600384153571

In [14]:
#of the Adaptive787 - 2.5% are shared with the PNP434 sequences sequences:
float(len(sharedsequencesBetweenCohorts))*100/len(countSharedAdaptive788samples)

2.497531522600647

In [17]:
print len(sharedsequencesBetweenCohorts[sharedsequencesBetweenCohorts['nSamples_PNP434']==1])

1513528


In [20]:
#78.2% of the sequences shared between cohorts, appear only once in the PNP434 cohort!

only1copy=sharedsequencesBetweenCohorts[sharedsequencesBetweenCohorts['nSamples_PNP434']==1]

percOnly1copyShared=float(len(only1copy))*100/len(sharedsequencesBetweenCohorts)
percOnly1copyShared

78.22933615611878

In [22]:
# 35.7% of all unique seqeucnes in the PNP434 cohort, are identified as private, but actually appear in the adaptive cohort as well:
notReallyPrivate=float(len(only1copy))*100/len(countSharedPNP434samples)
notReallyPrivate

35.66162818533619